### ISSUE TITLE
YOLO9 Training Problem

### ISSUE THREAD

**Issue by atunSensei** [Libre-YOLO/libreyolo#198]
First of all, I'm a complete noob. I learn thru osmosis, try-and-fail, and pattern recognition. But, I do not truly understand things like the pros do. So I apologize in advanced for my non-technical language. 

"My" background:
- I have managed to train a model using UL's YOLO.
- I have also managed to train a model using Baidu RT-DETR.
- Both using same dataset.
- I'm trying to replicate the above and get a working model using YOLO9 from LibreYOLO to get away from UL.

However, I have been encountering a training problem with LibreYOLO that I just cannot figure out. 

Background:
- I exported a CVAT dataset using UL Detection 1.0. Then I slice these for SAHI.
- I end up with images in "/images/train" and their exact filename.txt in "/labels/train"
- I have checked that the label coordinates "match" the objects I want to detect by manually drawing them, so as to check the SAHI was done correctly.

Problem:
- My model gets a 0 mAP. I see that the learning rate increases the first 3 epochs and then it stalls. The average loss drops the first three epochs and then it stalls. The end result when inferencing is completely wrong boxes with high confidence. I did notice though that the labels follow the same pattern as the labels I used to train, as if the model ended up memorizing/learning the relative position to each other, but never looked at a single image I gave it.
- Inference gets the label name correct and sizes correct. The model does indeed have the labels and boxes.

Things I've tried so far:
- I've checked the .png and .txt match in filename.
- I have checked that the label boxes in the labels/train .txt match the objects I want to detect by manually drawing them on their corresponding .png in the images/train folder.
- - I've modified data.yaml all the possible ways from straight as how I used it for UL to matching exactly how LibreYOLO describes the data.yaml (like indicating number of classes explicitly and then list the names as an array instead of 0: "name1", 1: "name2")
-  I've checked the .txt boxes to follow <class_id> <center_x> <center_y> <width> <height>

```
import os
from libreyolo import LibreYOLO9

def main():
    yaml_path = "data.yaml"
    model = LibreYOLO9("LibreYOLO9t.pt", size="t")

    results = model.train(
        data=yaml_path,          
        epochs=50,             
        imgsz=1024,               
        batch=8,                 
        device=0,               
        amp=True,                                        
        workers=8,               
        seed=0,
        cache=False,
        project="test",  
        name="run1",      
        exist_ok=True,           
        patience=20,             
        lr0=0.01,                
        optimizer="SGD"         
    )

if __name__ == '__main__':
    main()
```

At this point I'm questioning if I installed LibreYOLO right at all.

Any ideas? Thank you for any help.

**Comment by EHxuban11** [comment]
Hello @fegonzalezg , don't worry, let's figure out the solution together.

We've had an issue reporting that when the yaml contains incorrect anotations (wrong path of .txt or is using symlink), there is no warning or anything and it just trains and gets the map to 0. This has been fixed yesterday in main, but you may have an older version cloned. I recommend that you clone the repository and install main in editable mode to get this latest fixes.

And in general, if you could share the full contents of yout data.yaml, one sample txt, the training console output for the first epochs, that could be helpful. 

I think that the next immediate step for you is to just clone libreyolo and install the latest version in editable mode. Let me know if you get a warning saying that you have wrong labels.

**Comment by atunSensei** [comment]
Got no warnings. But I did see explicitly that it loaded annotations this time around. 

After 10 epochs I finally saw numbers different than 0 (pretty decent numbers actually). This time I used the .yalm in the same format as UL (which I tried many times before without success), so something got fixed independent of the data.yaml (?).

I'll let it run now and report back how the inference is doing. I'll report back.

Thank you so much.

**Comment by EHxuban11** [comment]
@fegonzalezg , happy to hear that, let me know if you need anything else :) 

If you manage to get nice results, you can close this issue.

**Comment by atunSensei** [comment]
@EHxuban11 so I ran inference, found huge offsets of the boxes. I was about to give up. But then I realized this "offset" bug occurred only when I passed an "imgsz" for inferencing. Docs do say one can bypass the default (640) by using imgsz, and since I trained at 1024, I naturally wanted to infer at 1024. But doing this breaks it. I """think""" the bug is in /models/base/inference.py, in 

```
def _predict_single(
        self,
        image: ImageInput,
        save: bool = False,
        output_path: str | None = None,
        conf: float = 0.25,
        iou: float = 0.45,
        imgsz: Optional[int] = None,
        classes: Optional[List[int]] = None,
        max_det: int = 300,
        color_format: str = "auto",
        output_file_format: Optional[str] = None,
        **kwargs,
    )
```

My little AI friend says that kwargs does not get the "imgsz" variable because it's explicitly written in the function's signature. Which I guess it doesn't matter because we latter resolve the size with:

`effective_imgsz = imgsz if imgsz is not None else self.model._get_input_size()`

Then we pass effective_imgsz  to preprocess, however, in Postprocess we are not passing effective_imgsz at all, assuming maybe it's inside kwargs? (which we fetch by using key "input_size" and not "imgsz", so it wouldn't work anyways?).

```
# Postprocess
        detections = self.model._postprocess(
            output, conf, iou, original_size, max_det=max_det, ratio=ratio, **kwargs
        )
```

Long story short 
- Passing an imgsz argument causes "Preprocess" to do its thing at whatever imgsize you gave it, but "Postprocess" defaults to 640. 
- NOT passing imgsz causes both "Postprocess" and "Preprocess" to fallback to the same default of 640, so it works (albeit at half the res in my case). I tested this out and indeed passing 640 DOES work.

**Comment by atunSensei** [comment]
@EHxuban11 also, ONNX Exporter leaves an "imgsz" in the metadata as "640". Feels like another hardcoded integer.
I should probably try learning how to contribute.

**Comment by EHxuban11** [comment]
Hey @fegonzalezg , 

Thanks for reporting these bugs, you are really using the library and finding real bugs which is exactly what I need this phase of the project.

I've investigated the bugs you report:

- BUG1: in inference the imgsz is not forwarded to postprocess so everything that is not 640 is not working.
- BUG2: ONNX export hardcodes 640 in metadata so potentially other sizes don't export correctly

Some info: we actively avoid to use hardcoded values because it makes the library hard to mantain and riddled with bugs. So nice catch

About next steps, you seem interested in contributing (and we'd love you to contribute), so if you want this can be your first contrib. We have a contributing guide here, you can take a look at it: https://github.com/LibreYOLO/libreyolo/blob/main/CONTRIBUTING.md

In your case we have already discussed the problem and so you can proceed to do a PR to main (we put bug fixes directly to main)

Now, about the solutions, Claude Code is telling me this, but feel free to not base the solution on this:

`
Bug1: In libreyolo/models/base/inference.py, forward input_size=effective_imgsz into every _postprocess call (currently passed only to _preprocess).

Bug2:  In libreyolo/export/exporter.py (_build_onnx_metadata and the TorchScript metadata builder), use the resolved export imgsz instead of self.model._get_input_size() when writing the "imgsz" metadata field.
`
